# Sentiment-based product recommendation system
This have two main modules.
1. Sentiment Analysis: We analyzed the user-review-text to check Positive or Negative review about product.
2. Recommendation system: This conains mainly recommendation based on Rating given to product. 

# Recommendation system

## 1 EDA


### 1.1 Import Lib and load data

In [1]:
# import libraties
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
# Reading review file
reviews = pd.read_csv('./dataset/sample30.csv')
print('Shape: ', reviews.shape)
reviews.head()

Shape:  (30000, 15)


id            brand  \
0  AV13O1A8GV-KLJ3akUyj  Universal Music   
1  AV14LG0R-jtxr-f38QfS         Lundberg   
2  AV14LG0R-jtxr-f38QfS         Lundberg   
3  AV16khLE-jtxr-f38VFn              K-Y   
4  AV16khLE-jtxr-f38VFn              K-Y   

                                          categories  \
0  Movies, Music & Books,Music,R&b,Movies & TV,Mo...   
1  Food,Packaged Foods,Snacks,Crackers,Snacks, Co...   
2  Food,Packaged Foods,Snacks,Crackers,Snacks, Co...   
3  Personal Care,Medicine Cabinet,Lubricant/Sperm...   
4  Personal Care,Medicine Cabinet,Lubricant/Sperm...   

                         manufacturer  \
0  Universal Music Group / Cash Money   
1                            Lundberg   
2                            Lundberg   
3                                 K-Y   
4                                 K-Y   

                                         name              reviews_date  \
0   Pink Friday: Roman Reloaded Re-Up (w/dvd)  2012-11-30T06:21:45.000Z   
1  Lundberg Organic Cinnamon Toast Rice Cakes  2017-07-09T00:00:00.000Z   
2  Lundberg Organic Cinnamon Toast Rice Cakes  2017-07-09T00:00:00.000Z   
3            K-Y Love Sensuality Pleasure Gel  2016-01-06T00:00:00.000Z   
4            K-Y Love Sensuality Pleasure Gel  2016-12-21T00:00:00.000Z   

  reviews_didPurchase reviews_doRecommend  reviews_rating  \
0                 NaN                 NaN               5   
1                True                 NaN               5   
2                True                 NaN               5   
3               False               False               1   
4               False               False               1   

                                        reviews_text reviews_title  \
0  i love this album. it's very good. more to the...  Just Awesome   
1  Good flavor. This review was collected as part...          Good   
2                                       Good flavor.          Good   
3  I read through the reviews on here before look...  Disappointed   
4  My husband bought this gel for us. The gel cau...    Irritation   

  reviews_userCity reviews_userProvince reviews_username user_sentiment  
0      Los Angeles                  NaN           joshua       Positive  
1              NaN                  NaN        dorothy w       Positive  
2              NaN                  NaN        dorothy w       Positive  
3              NaN                  NaN          rebecca       Negative  
4              NaN                  NaN        walker557       Negative

### Select only necessary columns only

In [4]:
df_reviews = reviews[['name','reviews_rating','reviews_username']]
print('Shape: ', df_reviews.shape)
df_reviews.head(2)

Shape:  (30000, 3)


name  reviews_rating reviews_username
0   Pink Friday: Roman Reloaded Re-Up (w/dvd)               5           joshua
1  Lundberg Organic Cinnamon Toast Rice Cakes               5        dorothy w

### There are few cases where we have duplicate rows and null values. Let's remove duplicate rows and null value
Removing duplicate rows is nessary for Pivot table.

In [5]:
df_reviews = df_reviews.drop_duplicates(subset=["name", "reviews_username"])

In [6]:
df_reviews.isnull().sum()

name                 0
reviews_rating       0
reviews_username    17
dtype: int64

In [7]:
# We have sufficent data and reviews_username is important field, so we can drop null data
df_reviews = df_reviews[~df_reviews.reviews_username.isnull()]
print('Shape:', df_reviews.shape)
df_reviews.isnull().sum()

Shape: (27588, 3)


name                0
reviews_rating      0
reviews_username    0
dtype: int64

We have now clean data and there are 27588 rows.

### Spliting the dataset into train and test

In [8]:
# Test and Train split of the dataset.
from sklearn.model_selection import train_test_split
train, test = train_test_split(df_reviews, test_size=0.30, random_state=31)

In [9]:
print(train.shape)
print(test.shape)

(19311, 3)
(8277, 3)


In [14]:
# Pivot the train ratings' dataset into matrix format in which columns are product name and the rows are reviews_username.
df_pivot = test.pivot(
    index='reviews_username',
    columns='name',
    values='reviews_rating'
).fillna(0)

df_pivot.head(3)

name                       100:Complete First Season (blu-Ray)  \
reviews_username                                                 
00sab00                                                    0.0   
10ten                                                      0.0   
11111111aaaaaaaaaaaaaaaaa                                  0.0   

name                       2017-2018 Brownline174 Duraflex 14-Month Planner 8 1/2 X 11 Black  \
reviews_username                                                                               
00sab00                                                                  0.0                   
10ten                                                                    0.0                   
11111111aaaaaaaaaaaaaaaaa                                                0.0                   

name                       2x Ultra Era with Oxi Booster, 50fl oz  \
reviews_username                                                    
00sab00                                                       0.0   
10ten                                                         0.0   
11111111aaaaaaaaaaaaaaaaa                                     0.0   

name                       4C Grated Parmesan Cheese 100% Natural 8oz Shaker  \
reviews_username                                                               
00sab00                                                                  0.0   
10ten                                                                    0.0   
11111111aaaaaaaaaaaaaaaaa                                                0.0   

name                       Africa's Best No-Lye Dual Conditioning Relaxer System Super  \
reviews_username                                                                         
00sab00                                                                  0.0             
10ten                                                                    0.0             
11111111aaaaaaaaaaaaaaaaa                                                0.0             

name                       Alex Cross (dvdvideo)  \
reviews_username                                   
00sab00                                      0.0   
10ten                                        0.0   
11111111aaaaaaaaaaaaaaaaa                    0.0   

name                       All,bran Complete Wheat Flakes, 18 Oz.  \
reviews_username                                                    
00sab00                                                       0.0   
10ten                                                         0.0   
11111111aaaaaaaaaaaaaaaaa                                     0.0   

name                       Annie's Homegrown Deluxe Elbows & Four Cheese Sauce  \
reviews_username                                                                 
00sab00                                                                  0.0     
10ten                                                                    0.0     
11111111aaaaaaaaaaaaaaaaa                                                0.0     

name                       Annie's Homegrown Gluten Free Double Chocolate Chip Granola Bars  \
reviews_username                                                                              
00sab00                                                                  0.0                  
10ten                                                                    0.0                  
11111111aaaaaaaaaaaaaaaaa                                                0.0                  

name                       Arrid Extra Dry Anti-Perspirant Deodorant Spray Regular  \
reviews_username                                                                     
00sab00                                                                  0.0         
10ten                                                                    0.0         
11111111aaaaaaaaaaaaaaaaa                                                0.0         

name                       ...  \
reviews_username           ... 

In [15]:
df_pivot.shape

(7948, 212)

In [16]:
# Copy the train dataset into dummy_train
dummy_train = train.copy()

In [17]:
# The product not rated by user is marked as 1 for prediction. 
dummy_train['reviews_rating'] = dummy_train['reviews_rating'].apply(lambda x: 0 if x>=1 else 1)

In [18]:
# Convert the dummy train dataset into matrix format.
dummy_train = dummy_train.pivot(
    index='reviews_username',
    columns='name',
    values='reviews_rating'
).fillna(1)

In [19]:
dummy_train.head()

name              0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  \
reviews_username                                                                  
00dog3                                                          1.0               
00sab00                                                         1.0               
01impala                                                        1.0               
02dakota                                                        1.0               
02deuce                                                         1.0               

name              100:Complete First Season (blu-Ray)  \
reviews_username                                        
00dog3                                            1.0   
00sab00                                           1.0   
01impala                                          1.0   
02dakota                                          1.0   
02deuce                                           1.0   

name              2x Ultra Era with Oxi Booster, 50fl oz  \
reviews_username                                           
00dog3                                               1.0   
00sab00                                              1.0   
01impala                                             1.0   
02dakota                                             1.0   
02deuce                                              1.0   

name              42 Dual Drop Leaf Table with 2 Madrid Chairs"  \
reviews_username                                                  
00dog3                                                      1.0   
00sab00                                                     1.0   
01impala                                                    1.0   
02dakota                                                    1.0   
02deuce                                                     1.0   

name              4C Grated Parmesan Cheese 100% Natural 8oz Shaker  \
reviews_username                                                      
00dog3                                                          1.0   
00sab00                                                         1.0   
01impala                                                        1.0   
02dakota                                                        1.0   
02deuce                                                         1.0   

name              5302050 15/16 FCT/HOSE ADAPTOR  \
reviews_username                                   
00dog3                                       1.0   
00sab00                                      1.0   
01impala                                     1.0   
02dakota                                     1.0   
02deuce                                      1.0   

name              Africa's Best No-Lye Dual Conditioning Relaxer System Super  \
reviews_username                                                                
00dog3                                                          1.0             
00sab00                                                         1.0             
01impala                                                        1.0             
02dakota                                                        1.0             
02deuce                                                         1.0             

name              Alberto VO5 Salon Series Smooth Plus Sleek Shampoo  \
reviews_username                                                       
00dog3                                                          1.0    
00sab00                                                         1.0    
01impala                                                        1.0    
02dakota                                                        1.0    
02deuce                                                         1.0    

name              Alex Cross (dvdvideo)  \
reviews_username                          
00dog3                              1.0   
00sab00                             1.0   
01impala           

In [20]:
dummy_train.shape

(17891, 255)

# Cosine Similarity

In [21]:
from sklearn.metrics.pairwise import pairwise_distances

# Creating the User Similarity Matrix using pairwise_distance function.
user_correlation = 1 - pairwise_distances(df_pivot, metric='cosine')
user_correlation[np.isnan(user_correlation)] = 0
print(user_correlation)

[[1. 0. 0. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]
 [0. 0. 1. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 1. 0. 0.]
 [0. 0. 0. ... 0. 1. 0.]
 [0. 0. 0. ... 0. 0. 1.]]


In [22]:
user_correlation.shape

(7948, 7948)

# Adjusted Cosine

In [23]:
# Create a user-product matrix.
df_pivot = train.pivot(
    index='reviews_username',
    columns='name',
    values='reviews_rating'
)

In [24]:
df_pivot.head()

name              0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  \
reviews_username                                                                  
00dog3                                                          NaN               
00sab00                                                         NaN               
01impala                                                        NaN               
02dakota                                                        NaN               
02deuce                                                         NaN               

name              100:Complete First Season (blu-Ray)  \
reviews_username                                        
00dog3                                            NaN   
00sab00                                           NaN   
01impala                                          NaN   
02dakota                                          NaN   
02deuce                                           NaN   

name              2x Ultra Era with Oxi Booster, 50fl oz  \
reviews_username                                           
00dog3                                               NaN   
00sab00                                              NaN   
01impala                                             NaN   
02dakota                                             NaN   
02deuce                                              NaN   

name              42 Dual Drop Leaf Table with 2 Madrid Chairs"  \
reviews_username                                                  
00dog3                                                      NaN   
00sab00                                                     NaN   
01impala                                                    NaN   
02dakota                                                    NaN   
02deuce                                                     NaN   

name              4C Grated Parmesan Cheese 100% Natural 8oz Shaker  \
reviews_username                                                      
00dog3                                                          NaN   
00sab00                                                         NaN   
01impala                                                        NaN   
02dakota                                                        NaN   
02deuce                                                         NaN   

name              5302050 15/16 FCT/HOSE ADAPTOR  \
reviews_username                                   
00dog3                                       NaN   
00sab00                                      NaN   
01impala                                     NaN   
02dakota                                     NaN   
02deuce                                      NaN   

name              Africa's Best No-Lye Dual Conditioning Relaxer System Super  \
reviews_username                                                                
00dog3                                                          NaN             
00sab00                                                         NaN             
01impala                                                        NaN             
02dakota                                                        NaN             
02deuce                                                         NaN             

name              Alberto VO5 Salon Series Smooth Plus Sleek Shampoo  \
reviews_username                                                       
00dog3                                                          NaN    
00sab00                                                         NaN    
01impala                                                        NaN    
02dakota                                                        NaN    
02deuce                                                         NaN    

name              Alex Cross (dvdvideo)  \
reviews_username                          
00dog3                              NaN   
00sab00                             NaN   
01impala           

### Normalising the rating of the product for each user around 0 mean

In [25]:
mean = np.nanmean(df_pivot, axis=1)
df_subtracted = (df_pivot.T-mean).T
df_subtracted.head()

name              0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  \
reviews_username                                                                  
00dog3                                                          NaN               
00sab00                                                         NaN               
01impala                                                        NaN               
02dakota                                                        NaN               
02deuce                                                         NaN               

name              100:Complete First Season (blu-Ray)  \
reviews_username                                        
00dog3                                            NaN   
00sab00                                           NaN   
01impala                                          NaN   
02dakota                                          NaN   
02deuce                                           NaN   

name              2x Ultra Era with Oxi Booster, 50fl oz  \
reviews_username                                           
00dog3                                               NaN   
00sab00                                              NaN   
01impala                                             NaN   
02dakota                                             NaN   
02deuce                                              NaN   

name              42 Dual Drop Leaf Table with 2 Madrid Chairs"  \
reviews_username                                                  
00dog3                                                      NaN   
00sab00                                                     NaN   
01impala                                                    NaN   
02dakota                                                    NaN   
02deuce                                                     NaN   

name              4C Grated Parmesan Cheese 100% Natural 8oz Shaker  \
reviews_username                                                      
00dog3                                                          NaN   
00sab00                                                         NaN   
01impala                                                        NaN   
02dakota                                                        NaN   
02deuce                                                         NaN   

name              5302050 15/16 FCT/HOSE ADAPTOR  \
reviews_username                                   
00dog3                                       NaN   
00sab00                                      NaN   
01impala                                     NaN   
02dakota                                     NaN   
02deuce                                      NaN   

name              Africa's Best No-Lye Dual Conditioning Relaxer System Super  \
reviews_username                                                                
00dog3                                                          NaN             
00sab00                                                         NaN             
01impala                                                        NaN             
02dakota                                                        NaN             
02deuce                                                         NaN             

name              Alberto VO5 Salon Series Smooth Plus Sleek Shampoo  \
reviews_username                                                       
00dog3                                                          NaN    
00sab00                                                         NaN    
01impala                                                        NaN    
02dakota                                                        NaN    
02deuce                                                         NaN    

name              Alex Cross (dvdvideo)  \
reviews_username                          
00dog3                              NaN   
00sab00                             NaN   
01impala           

### Finding cosine similarity

In [26]:
from sklearn.metrics.pairwise import pairwise_distances

In [27]:
# Creating the User Similarity Matrix using pairwise_distance function.
user_correlation = 1 - pairwise_distances(df_subtracted.fillna(0), metric='cosine')
user_correlation[np.isnan(user_correlation)] = 0
print(user_correlation)

[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]


## Prediction - User User

In [28]:
user_correlation[user_correlation<0]=0
user_correlation

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [29]:
user_predicted_ratings = np.dot(user_correlation, df_pivot.fillna(0))
user_predicted_ratings

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [30]:
user_predicted_ratings.shape

(17891, 255)

In [31]:
# Not Rated users
user_final_rating = np.multiply(user_predicted_ratings,dummy_train)
user_final_rating.head()

name              0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  \
reviews_username                                                                  
00dog3                                                          0.0               
00sab00                                                         0.0               
01impala                                                        0.0               
02dakota                                                        0.0               
02deuce                                                         0.0               

name              100:Complete First Season (blu-Ray)  \
reviews_username                                        
00dog3                                            0.0   
00sab00                                           0.0   
01impala                                          0.0   
02dakota                                          0.0   
02deuce                                           0.0   

name              2x Ultra Era with Oxi Booster, 50fl oz  \
reviews_username                                           
00dog3                                               0.0   
00sab00                                              0.0   
01impala                                             0.0   
02dakota                                             0.0   
02deuce                                              0.0   

name              42 Dual Drop Leaf Table with 2 Madrid Chairs"  \
reviews_username                                                  
00dog3                                                      0.0   
00sab00                                                     0.0   
01impala                                                    0.0   
02dakota                                                    0.0   
02deuce                                                     0.0   

name              4C Grated Parmesan Cheese 100% Natural 8oz Shaker  \
reviews_username                                                      
00dog3                                                          0.0   
00sab00                                                         0.0   
01impala                                                        0.0   
02dakota                                                        0.0   
02deuce                                                         0.0   

name              5302050 15/16 FCT/HOSE ADAPTOR  \
reviews_username                                   
00dog3                                       0.0   
00sab00                                      0.0   
01impala                                     0.0   
02dakota                                     0.0   
02deuce                                      0.0   

name              Africa's Best No-Lye Dual Conditioning Relaxer System Super  \
reviews_username                                                                
00dog3                                                          0.0             
00sab00                                                         0.0             
01impala                                                        0.0             
02dakota                                                        0.0             
02deuce                                                         0.0             

name              Alberto VO5 Salon Series Smooth Plus Sleek Shampoo  \
reviews_username                                                       
00dog3                                                          0.0    
00sab00                                                         0.0    
01impala                                                        0.0    
02dakota                                                        0.0    
02deuce                                                         0.0    

name              Alex Cross (dvdvideo)  \
reviews_username                          
00dog3                              0.0   
00sab00                             0.0   
01impala           

In [32]:
# Take the user ID as input.
# user_input = input("Enter your user name")
# print(user_input)
# d = user_final_rating.loc[user_input].sort_values(ascending=False)[0:5]
# d = pd.merge(d,reviews,left_on='name',right_on='name', how = 'left')
# d.head()

# Evaluation - User User

In [33]:
# Find out the common users of test and train dataset.
common = test[test.reviews_username.isin(train.reviews_username)]
common.shape

(1106, 3)

In [34]:
common.head()

name  reviews_rating  \
4245   Hoover174 Platinum Collection153 Lightweight B...               5   
19882               Clorox Disinfecting Bathroom Cleaner               5   
26308  Stargate (ws) (ultimate Edition) (director's C...               4   
6284                              Hormel Chili, No Beans               3   
15580                      D-Con Mice Bait Station - 3ct               5   

      reviews_username  
4245            teresa  
19882            shaba  
26308           jayjay  
6284              meme  
15580             nana

In [35]:
# convert into the user-based matrix.
common_user_based_matrix = common.pivot_table(index='reviews_username', columns='name', values='reviews_rating')

In [36]:
# Convert the user_correlation matrix into dataframe.
user_correlation_df = pd.DataFrame(user_correlation)

In [37]:
user_correlation_df['reviews_username'] = df_subtracted.index
user_correlation_df.set_index('reviews_username',inplace=True)
user_correlation_df.head()

0      1      2      3      4      5      6      7      \
reviews_username                                                           
00dog3              0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
00sab00             0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
01impala            0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
02dakota            0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
02deuce             0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

                  8      9      ...  17881  17882  17883  17884  17885  17886  \
reviews_username                ...                                             
00dog3              0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0   
00sab00             0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0   
01impala            0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0   
02dakota            0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0   
02deuce             0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0   

                  17887  17888  17889  17890  
reviews_username                              
00dog3              0.0    0.0    0.0    0.0  
00sab00             0.0    0.0    0.0    0.0  
01impala            0.0    0.0    0.0    0.0  
02dakota            0.0    0.0    0.0    0.0  
02deuce             0.0    0.0    0.0    0.0  

[5 rows x 17891 columns]

In [38]:
common.head(1)

name  reviews_rating  \
4245  Hoover174 Platinum Collection153 Lightweight B...               5   

     reviews_username  
4245           teresa

In [39]:
user_correlation_df.columns

Index([    0,     1,     2,     3,     4,     5,     6,     7,     8,     9,
       ...
       17881, 17882, 17883, 17884, 17885, 17886, 17887, 17888, 17889, 17890],
      dtype='object', length=17891)

In [40]:
user_correlation_df.head()

0      1      2      3      4      5      6      7      \
reviews_username                                                           
00dog3              0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
00sab00             0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
01impala            0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
02dakota            0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
02deuce             0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

                  8      9      ...  17881  17882  17883  17884  17885  17886  \
reviews_username                ...                                             
00dog3              0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0   
00sab00             0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0   
01impala            0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0   
02dakota            0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0   
02deuce             0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0   

                  17887  17888  17889  17890  
reviews_username                              
00dog3              0.0    0.0    0.0    0.0  
00sab00             0.0    0.0    0.0    0.0  
01impala            0.0    0.0    0.0    0.0  
02dakota            0.0    0.0    0.0    0.0  
02deuce             0.0    0.0    0.0    0.0  

[5 rows x 17891 columns]

In [41]:
list_name = common.reviews_username.tolist()
user_correlation_df.columns = df_subtracted.index.tolist()
user_correlation_df_1 =  user_correlation_df[user_correlation_df.index.isin(list_name)]
print('Shape: ',user_correlation_df_1.shape)
user_correlation_df_1.head()

Shape:  (925, 17891)


00dog3  00sab00  01impala  02dakota  02deuce  0325home  \
reviews_username                                                           
00sab00              0.0      0.0       0.0       0.0      0.0       0.0   
1234                 0.0      0.0       0.0       0.0      0.0       0.0   
1943                 0.0      0.0       0.0       0.0      0.0       0.0   
37f5p                0.0      0.0       0.0       0.0      0.0       0.0   
50cal                0.0      0.0       0.0       0.0      0.0       0.0   

                  06stidriver  08dallas  09mommy11  1.11E+24  ...  zotox  \
reviews_username                                              ...          
00sab00                   0.0       0.0        0.0       0.0  ...    0.0   
1234                      0.0       0.0        0.0       0.0  ...    0.0   
1943                      0.0       0.0        0.0       0.0  ...    0.0   
37f5p                     0.0       0.0        0.0       0.0  ...    0.0   
50cal                     0.0       0.0        0.0       0.0  ...    0.0   

                  zout22389  zsarah  zsazsa  zulaa118  zwithanx  zxcsdfd  \
reviews_username                                                           
00sab00                 0.0     0.0     0.0       0.0       0.0      0.0   
1234                    0.0     0.0     0.0       0.0       0.0      0.0   
1943                    0.0     0.0     0.0       0.0       0.0      0.0   
37f5p                   0.0     0.0     0.0       0.0       0.0      0.0   
50cal                   0.0     0.0     0.0       0.0       0.0      0.0   

                  zxjki  zzdiane  zzz1127  
reviews_username                           
00sab00             0.0      0.0      0.0  
1234                0.0      0.0      0.0  
1943                0.0      0.0      0.0  
37f5p               0.0      0.0      0.0  
50cal               0.0      0.0      0.0  

[5 rows x 17891 columns]

In [42]:
user_correlation_df_2 = user_correlation_df_1.T[user_correlation_df_1.T.index.isin(list_name)]
user_correlation_df_3 = user_correlation_df_2.T
print('Shape: ', user_correlation_df_3.shape)
user_correlation_df_3.head()

Shape:  (925, 925)


00sab00  1234  1943  37f5p  50cal  aaron  abbey  abby  \
reviews_username                                                          
00sab00               0.0   0.0   0.0    0.0    0.0    0.0    0.0   0.0   
1234                  0.0   0.0   0.0    0.0    0.0    0.0    0.0   0.0   
1943                  0.0   0.0   0.0    0.0    0.0    0.0    0.0   0.0   
37f5p                 0.0   0.0   0.0    0.0    0.0    0.0    0.0   0.0   
50cal                 0.0   0.0   0.0    0.0    0.0    0.0    0.0   0.0   

                  abismomy  acv4217  ...  wil27  wildman  willy  wimbly  \
reviews_username                     ...                                  
00sab00                0.0      0.0  ...    0.0      0.0    0.0     0.0   
1234                   0.0      0.0  ...    0.0      0.0    0.0     0.0   
1943                   0.0      0.0  ...    0.0      0.0    0.0     0.0   
37f5p                  0.0      0.0  ...    0.0      0.0    0.0     0.0   
50cal                  0.0      0.0  ...    0.0      0.0    0.0     0.0   

                  wolverine  xmom  yeyo  yummy  zippy  zitro  
reviews_username                                              
00sab00                 0.0   0.0   0.0    0.0    0.0    0.0  
1234                    0.0   0.0   0.0    0.0    0.0    0.0  
1943                    0.0   0.0   0.0    0.0    0.0    0.0  
37f5p                   0.0   0.0   0.0    0.0    0.0    0.0  
50cal                   0.0   0.0   0.0    0.0    0.0    0.0  

[5 rows x 925 columns]

In [43]:
user_correlation_df_3.shape

(925, 925)

In [44]:
user_correlation_df_3[user_correlation_df_3<0]=0

common_user_predicted_ratings = np.dot(user_correlation_df_3, common_user_based_matrix.fillna(0))
common_user_predicted_ratings

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [45]:
dummy_test = common.copy()

dummy_test['reviews_rating'] = dummy_test['reviews_rating'].apply(lambda x: 1 if x>=1 else 0)

dummy_test = dummy_test.pivot_table(
    
    index='reviews_username',
    columns='name',
    values='reviews_rating'
).fillna(0)

In [46]:
dummy_test.shape

(925, 126)

In [47]:
common_user_predicted_ratings = np.multiply(common_user_predicted_ratings,dummy_test)
common_user_predicted_ratings.head(2)                  

name              100:Complete First Season (blu-Ray)  Alex Cross (dvdvideo)  \
reviews_username                                                               
00sab00                                           0.0                    0.0   
1234                                              0.0                    0.0   

name              All,bran Complete Wheat Flakes, 18 Oz.  \
reviews_username                                           
00sab00                                              0.0   
1234                                                 0.0   

name              Aussie Aussome Volume Shampoo, 13.5 Oz  \
reviews_username                                           
00sab00                                              0.0   
1234                                                 0.0   

name              Australian Gold Exotic Blend Lotion, SPF 4  \
reviews_username                                               
00sab00                                                  0.0   
1234                                                     0.0   

name              Aveeno Baby Continuous Protection Lotion Sunscreen with Broad Spectrum SPF 55, 4oz  \
reviews_username                                                                                       
00sab00                                                         0.0                                    
1234                                                            0.0                                    

name              Avery174 Ready Index Contemporary Table Of Contents Divider, 1-8, Multi, Letter  \
reviews_username                                                                                    
00sab00                                                         0.0                                 
1234                                                            0.0                                 

name              Banana Boat Sunless Summer Color Self Tanning Lotion, Light To Medium  \
reviews_username                                                                          
00sab00                                                         0.0                       
1234                                                            0.0                       

name              Barielle Nail Rebuilding Protein  \
reviews_username                                     
00sab00                                        0.0   
1234                                           0.0   

name              Ben & Jerry's Coffee, Coffee Buzzbuzzbuzz! Ice Cream, Pint  \
reviews_username                                                               
00sab00                                                         0.0            
1234                                                            0.0            

name              ...  Various - Red Hot Blue:Tribute To Cole Porter (cd)  \
reviews_username  ...                                                       
00sab00           ...                                                0.0    
1234              ...                                                0.0    

name              Various Artists - Choo Choo Soul (cd)  \
reviews_username                                          
00sab00                                             0.0   
1234                                                0.0   

name              Vaseline Intensive Care Healthy Hands Stronger Nails  \
reviews_username                                                         
00sab00                                                         0.0      
1234                                                            0.0      

name              Vaseline Intensive Care Lip Therapy Cocoa Butter  \
reviews_username                                                     
00sab00                                                        0.0   
1234                                                           0.0   

name              Vicks Vaporub, Regular, 3.53oz  \
reviews_username                       

In [48]:
common_user_predicted_ratings = np.multiply(common_user_predicted_ratings,dummy_test)
common_user_predicted_ratings.head(2)

name              100:Complete First Season (blu-Ray)  Alex Cross (dvdvideo)  \
reviews_username                                                               
00sab00                                           0.0                    0.0   
1234                                              0.0                    0.0   

name              All,bran Complete Wheat Flakes, 18 Oz.  \
reviews_username                                           
00sab00                                              0.0   
1234                                                 0.0   

name              Aussie Aussome Volume Shampoo, 13.5 Oz  \
reviews_username                                           
00sab00                                              0.0   
1234                                                 0.0   

name              Australian Gold Exotic Blend Lotion, SPF 4  \
reviews_username                                               
00sab00                                                  0.0   
1234                                                     0.0   

name              Aveeno Baby Continuous Protection Lotion Sunscreen with Broad Spectrum SPF 55, 4oz  \
reviews_username                                                                                       
00sab00                                                         0.0                                    
1234                                                            0.0                                    

name              Avery174 Ready Index Contemporary Table Of Contents Divider, 1-8, Multi, Letter  \
reviews_username                                                                                    
00sab00                                                         0.0                                 
1234                                                            0.0                                 

name              Banana Boat Sunless Summer Color Self Tanning Lotion, Light To Medium  \
reviews_username                                                                          
00sab00                                                         0.0                       
1234                                                            0.0                       

name              Barielle Nail Rebuilding Protein  \
reviews_username                                     
00sab00                                        0.0   
1234                                           0.0   

name              Ben & Jerry's Coffee, Coffee Buzzbuzzbuzz! Ice Cream, Pint  \
reviews_username                                                               
00sab00                                                         0.0            
1234                                                            0.0            

name              ...  Various - Red Hot Blue:Tribute To Cole Porter (cd)  \
reviews_username  ...                                                       
00sab00           ...                                                0.0    
1234              ...                                                0.0    

name              Various Artists - Choo Choo Soul (cd)  \
reviews_username                                          
00sab00                                             0.0   
1234                                                0.0   

name              Vaseline Intensive Care Healthy Hands Stronger Nails  \
reviews_username                                                         
00sab00                                                         0.0      
1234                                                            0.0      

name              Vaseline Intensive Care Lip Therapy Cocoa Butter  \
reviews_username                                                     
00sab00                                                        0.0   
1234                                                           0.0   

name              Vicks Vaporub, Regular, 3.53oz  \
reviews_username                       

In [49]:
from sklearn.preprocessing import MinMaxScaler
from numpy import *

X  = common_user_predicted_ratings.copy() 
X = X[X>0]

scaler = MinMaxScaler(feature_range=(1, 5))
print(scaler.fit(X))
y = (scaler.transform(X))

print(y)

MinMaxScaler(feature_range=(1, 5))
[[nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 ...
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]]


C:\Users\BA06878\Anaconda3\lib\site-packages\sklearn\preprocessing\_data.py:461: RuntimeWarning: All-NaN slice encountered
  data_min = np.nanmin(X, axis=0)
C:\Users\BA06878\Anaconda3\lib\site-packages\sklearn\preprocessing\_data.py:462: RuntimeWarning: All-NaN slice encountered
  data_max = np.nanmax(X, axis=0)


In [50]:
common_ = common.pivot_table(
    index='reviews_username',
    columns='name',
    values='reviews_rating'
)
common_.shape

(925, 126)

In [51]:
# Finding total non-NaN value
total_non_nan = np.count_nonzero(~np.isnan(y))

In [52]:
rmse = (sum(sum((common_ - y )**2))/total_non_nan)**0.5
print(rmse)

2.5597195521751623


# Item Based Similarity

In [53]:
df_pivot = train.pivot(
    index='reviews_username',
    columns='name',
    values='reviews_rating'
).T

df_pivot.head()

reviews_username                                    00dog3  00sab00  01impala  \
name                                                                            
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...     NaN      NaN       NaN   
100:Complete First Season (blu-Ray)                    NaN      NaN       NaN   
2x Ultra Era with Oxi Booster, 50fl oz                 NaN      NaN       NaN   
42 Dual Drop Leaf Table with 2 Madrid Chairs"          NaN      NaN       NaN   
4C Grated Parmesan Cheese 100% Natural 8oz Shaker      NaN      NaN       NaN   

reviews_username                                    02dakota  02deuce  \
name                                                                    
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...       NaN      NaN   
100:Complete First Season (blu-Ray)                      NaN      NaN   
2x Ultra Era with Oxi Booster, 50fl oz                   NaN      NaN   
42 Dual Drop Leaf Table with 2 Madrid Chairs"            NaN      NaN   
4C Grated Parmesan Cheese 100% Natural 8oz Shaker        NaN      NaN   

reviews_username                                    0325home  06stidriver  \
name                                                                        
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...       NaN          NaN   
100:Complete First Season (blu-Ray)                      NaN          NaN   
2x Ultra Era with Oxi Booster, 50fl oz                   NaN          NaN   
42 Dual Drop Leaf Table with 2 Madrid Chairs"            NaN          NaN   
4C Grated Parmesan Cheese 100% Natural 8oz Shaker        NaN          NaN   

reviews_username                                    08dallas  09mommy11  \
name                                                                      
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...       NaN        NaN   
100:Complete First Season (blu-Ray)                      5.0        NaN   
2x Ultra Era with Oxi Booster, 50fl oz                   NaN        NaN   
42 Dual Drop Leaf Table with 2 Madrid Chairs"            NaN        NaN   
4C Grated Parmesan Cheese 100% Natural 8oz Shaker        NaN        NaN   

reviews_username                                    1.11E+24  ...  zotox  \
name                                                          ...          
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...       NaN  ...    NaN   
100:Complete First Season (blu-Ray)                      NaN  ...    NaN   
2x Ultra Era with Oxi Booster, 50fl oz                   NaN  ...    NaN   
42 Dual Drop Leaf Table with 2 Madrid Chairs"            NaN  ...    NaN   
4C Grated Parmesan Cheese 100% Natural 8oz Shaker        NaN  ...    NaN   

reviews_username                                    zout22389  zsarah  zsazsa  \
name                                                                            
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...        NaN     NaN     NaN   
100:Complete First Season (blu-Ray)                       NaN     NaN     NaN   
2x Ultra Era with Oxi Booster, 50fl oz                    NaN     NaN     NaN   
42 Dual Drop Leaf Table with 2 Madrid Chairs"             NaN     NaN     NaN   
4C Grated Parmesan Cheese 100% Natural 8oz Shaker         NaN     NaN     NaN   

reviews_username                                    zulaa118  zwithanx  \
name                                                                     
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...       NaN       NaN   
100:Complete First Season (blu-Ray)                      NaN       NaN   
2x Ultra Era with Oxi Booster, 50fl oz                   NaN       NaN   
42 Dual Drop Leaf Table with 2 Madrid Chairs"            NaN       NaN   
4C Grated Parmesan Cheese 100% Natural 8oz Shaker        NaN       NaN   

reviews_username                                    zxcsdfd  zxjki  zzdiane  \
name                                                                          
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...      NaN    NaN      NaN 

In [54]:
mean = np.nanmean(df_pivot, axis=1)
df_subtracted = (df_pivot.T-mean).T
df_subtracted.head()

reviews_username                                    00dog3  00sab00  01impala  \
name                                                                            
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...     NaN      NaN       NaN   
100:Complete First Season (blu-Ray)                    NaN      NaN       NaN   
2x Ultra Era with Oxi Booster, 50fl oz                 NaN      NaN       NaN   
42 Dual Drop Leaf Table with 2 Madrid Chairs"          NaN      NaN       NaN   
4C Grated Parmesan Cheese 100% Natural 8oz Shaker      NaN      NaN       NaN   

reviews_username                                    02dakota  02deuce  \
name                                                                    
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...       NaN      NaN   
100:Complete First Season (blu-Ray)                      NaN      NaN   
2x Ultra Era with Oxi Booster, 50fl oz                   NaN      NaN   
42 Dual Drop Leaf Table with 2 Madrid Chairs"            NaN      NaN   
4C Grated Parmesan Cheese 100% Natural 8oz Shaker        NaN      NaN   

reviews_username                                    0325home  06stidriver  \
name                                                                        
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...       NaN          NaN   
100:Complete First Season (blu-Ray)                      NaN          NaN   
2x Ultra Era with Oxi Booster, 50fl oz                   NaN          NaN   
42 Dual Drop Leaf Table with 2 Madrid Chairs"            NaN          NaN   
4C Grated Parmesan Cheese 100% Natural 8oz Shaker        NaN          NaN   

reviews_username                                    08dallas  09mommy11  \
name                                                                      
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...       NaN        NaN   
100:Complete First Season (blu-Ray)                 0.298969        NaN   
2x Ultra Era with Oxi Booster, 50fl oz                   NaN        NaN   
42 Dual Drop Leaf Table with 2 Madrid Chairs"            NaN        NaN   
4C Grated Parmesan Cheese 100% Natural 8oz Shaker        NaN        NaN   

reviews_username                                    1.11E+24  ...  zotox  \
name                                                          ...          
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...       NaN  ...    NaN   
100:Complete First Season (blu-Ray)                      NaN  ...    NaN   
2x Ultra Era with Oxi Booster, 50fl oz                   NaN  ...    NaN   
42 Dual Drop Leaf Table with 2 Madrid Chairs"            NaN  ...    NaN   
4C Grated Parmesan Cheese 100% Natural 8oz Shaker        NaN  ...    NaN   

reviews_username                                    zout22389  zsarah  zsazsa  \
name                                                                            
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...        NaN     NaN     NaN   
100:Complete First Season (blu-Ray)                       NaN     NaN     NaN   
2x Ultra Era with Oxi Booster, 50fl oz                    NaN     NaN     NaN   
42 Dual Drop Leaf Table with 2 Madrid Chairs"             NaN     NaN     NaN   
4C Grated Parmesan Cheese 100% Natural 8oz Shaker         NaN     NaN     NaN   

reviews_username                                    zulaa118  zwithanx  \
name                                                                     
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...       NaN       NaN   
100:Complete First Season (blu-Ray)                      NaN       NaN   
2x Ultra Era with Oxi Booster, 50fl oz                   NaN       NaN   
42 Dual Drop Leaf Table with 2 Madrid Chairs"            NaN       NaN   
4C Grated Parmesan Cheese 100% Natural 8oz Shaker        NaN       NaN   

reviews_username                                    zxcsdfd  zxjki  zzdiane  \
name                                                                          
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...      NaN    NaN      NaN 

In [55]:
from sklearn.metrics.pairwise import pairwise_distances

In [56]:
# Item Similarity Matrix
item_correlation = 1 - pairwise_distances(df_subtracted.fillna(0), metric='cosine')
item_correlation[np.isnan(item_correlation)] = 0
print(item_correlation)

[[1. 0. 0. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]
 [0. 0. 1. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 1. 0. 0.]
 [0. 0. 0. ... 0. 1. 0.]
 [0. 0. 0. ... 0. 0. 1.]]


In [57]:
item_correlation[item_correlation<0]=0
item_correlation

array([[1., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]])

# Prediction - Item Item

In [58]:
item_predicted_ratings = np.dot((df_pivot.fillna(0).T),item_correlation)
print('Shape: ', item_predicted_ratings.shape)
item_predicted_ratings

Shape:  (17891, 255)


array([[0.        , 0.        , 0.        , ..., 0.01928584, 0.        ,
        0.        ],
       [0.        , 0.01703714, 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.0450749 , 0.        , ..., 0.00914624, 0.00217527,
        0.01132432],
       ...,
       [0.        , 0.        , 0.        , ..., 0.00012819, 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.01333037,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.00025637, 0.        ,
        0.        ]])

In [59]:
dummy_train.shape

(17891, 255)

In [60]:
item_final_rating = np.multiply(item_predicted_ratings,dummy_train)
item_final_rating.head()

name              0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  \
reviews_username                                                                  
00dog3                                                          0.0               
00sab00                                                         0.0               
01impala                                                        0.0               
02dakota                                                        0.0               
02deuce                                                         0.0               

name              100:Complete First Season (blu-Ray)  \
reviews_username                                        
00dog3                                       0.000000   
00sab00                                      0.017037   
01impala                                     0.045075   
02dakota                                     0.075125   
02deuce                                      0.060100   

name              2x Ultra Era with Oxi Booster, 50fl oz  \
reviews_username                                           
00dog3                                               0.0   
00sab00                                              0.0   
01impala                                             0.0   
02dakota                                             0.0   
02deuce                                              0.0   

name              42 Dual Drop Leaf Table with 2 Madrid Chairs"  \
reviews_username                                                  
00dog3                                                      0.0   
00sab00                                                     0.0   
01impala                                                    0.0   
02dakota                                                    0.0   
02deuce                                                     0.0   

name              4C Grated Parmesan Cheese 100% Natural 8oz Shaker  \
reviews_username                                                      
00dog3                                                          0.0   
00sab00                                                         0.0   
01impala                                                        0.0   
02dakota                                                        0.0   
02deuce                                                         0.0   

name              5302050 15/16 FCT/HOSE ADAPTOR  \
reviews_username                                   
00dog3                                       0.0   
00sab00                                      0.0   
01impala                                     0.0   
02dakota                                     0.0   
02deuce                                      0.0   

name              Africa's Best No-Lye Dual Conditioning Relaxer System Super  \
reviews_username                                                                
00dog3                                                          0.0             
00sab00                                                         0.0             
01impala                                                        0.0             
02dakota                                                        0.0             
02deuce                                                         0.0             

name              Alberto VO5 Salon Series Smooth Plus Sleek Shampoo  \
reviews_username                                                       
00dog3                                                          0.0    
00sab00                                                         0.0    
01impala                                                        0.0    
02dakota                                                        0.0    
02deuce                                                         0.0    

name              Alex Cross (dvdvideo)  \
reviews_username                          
00dog3                          0.00000   
00sab00                         0.00293   
01impala           

### Finding the top 20 recommendation for the *user*

In [61]:
# Take the user ID as input
# user_input = input("Enter your user id")
# print(user_input)
# d = item_final_rating.loc[user_input].sort_values(ascending=False)[0:20]
# d = pd.merge(d,reviews,left_on='name',right_on='name', how = 'left')
# d.head()

# Evaluation - Item Item

In [62]:
common =  test[test.name.isin(train.name)]
print('Shape:',common.shape)
common.head()

Shape: (8253, 3)


name  reviews_rating  \
22245                              Alex Cross (dvdvideo)               4   
23511  Godzilla 3d Includes Digital Copy Ultraviolet ...               4   
7965   Clorox Disinfecting Wipes Value Pack Scented 1...               5   
4245   Hoover174 Platinum Collection153 Lightweight B...               5   
11201  Clorox Disinfecting Wipes Value Pack Scented 1...               5   

      reviews_username  
22245    albertacanada  
23511           reeree  
7965           adash75  
4245            teresa  
11201       zhynellebb

In [63]:
item_correlation_df = pd.DataFrame(item_correlation)
item_correlation_df.head(2)

0    1    2    3    4    5    6    7    8    9    ...  245  246  247  248  \
0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
1  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   

   249  250  251  252  253  254  
0  0.0  0.0  0.0  0.0  0.0  0.0  
1  0.0  0.0  0.0  0.0  0.0  0.0  

[2 rows x 255 columns]

In [64]:
item_correlation_df['name'] = df_subtracted.index
item_correlation_df.set_index('name',inplace=True)
item_correlation_df.head()

0    1    2    3    4    \
name                                                                          
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...  1.0  0.0  0.0  0.0  0.0   
100:Complete First Season (blu-Ray)                 0.0  1.0  0.0  0.0  0.0   
2x Ultra Era with Oxi Booster, 50fl oz              0.0  0.0  1.0  0.0  0.0   
42 Dual Drop Leaf Table with 2 Madrid Chairs"       0.0  0.0  0.0  0.0  0.0   
4C Grated Parmesan Cheese 100% Natural 8oz Shaker   0.0  0.0  0.0  0.0  0.0   

                                                    5    6    7    8    9    \
name                                                                          
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...  0.0  0.0  0.0  0.0  0.0   
100:Complete First Season (blu-Ray)                 0.0  0.0  0.0  0.0  0.0   
2x Ultra Era with Oxi Booster, 50fl oz              0.0  0.0  0.0  0.0  0.0   
42 Dual Drop Leaf Table with 2 Madrid Chairs"       0.0  0.0  0.0  0.0  0.0   
4C Grated Parmesan Cheese 100% Natural 8oz Shaker   0.0  0.0  0.0  0.0  0.0   

                                                    ...  245  246  247  248  \
name                                                ...                       
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...  ...  0.0  0.0  0.0  0.0   
100:Complete First Season (blu-Ray)                 ...  0.0  0.0  0.0  0.0   
2x Ultra Era with Oxi Booster, 50fl oz              ...  0.0  0.0  0.0  0.0   
42 Dual Drop Leaf Table with 2 Madrid Chairs"       ...  0.0  0.0  0.0  0.0   
4C Grated Parmesan Cheese 100% Natural 8oz Shaker   ...  0.0  0.0  0.0  0.0   

                                                    249  250  251  252  253  \
name                                                                          
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...  0.0  0.0  0.0  0.0  0.0   
100:Complete First Season (blu-Ray)                 0.0  0.0  0.0  0.0  0.0   
2x Ultra Era with Oxi Booster, 50fl oz              0.0  0.0  0.0  0.0  0.0   
42 Dual Drop Leaf Table with 2 Madrid Chairs"       0.0  0.0  0.0  0.0  0.0   
4C Grated Parmesan Cheese 100% Natural 8oz Shaker   0.0  0.0  0.0  0.0  0.0   

                                                    254  
name                                                     
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...  0.0  
100:Complete First Season (blu-Ray)                 0.0  
2x Ultra Era with Oxi Booster, 50fl oz              0.0  
42 Dual Drop Leaf Table with 2 Madrid Chairs"       0.0  
4C Grated Parmesan Cheese 100% Natural 8oz Shaker   0.0  

[5 rows x 255 columns]

In [65]:
list_name = common.name.tolist()
item_correlation_df.columns = df_subtracted.index.tolist()
item_correlation_df_1 =  item_correlation_df[item_correlation_df.index.isin(list_name)]

In [66]:
item_correlation_df_2 = item_correlation_df_1.T[item_correlation_df_1.T.index.isin(list_name)]
item_correlation_df_3 = item_correlation_df_2.T
item_correlation_df_3.head()

100:Complete First Season (blu-Ray)  \
name                                                                                      
100:Complete First Season (blu-Ray)                                                 1.0   
2x Ultra Era with Oxi Booster, 50fl oz                                              0.0   
4C Grated Parmesan Cheese 100% Natural 8oz Shaker                                   0.0   
Africa's Best No-Lye Dual Conditioning Relaxer ...                                  0.0   
Alex Cross (dvdvideo)                                                               0.0   

                                                    2x Ultra Era with Oxi Booster, 50fl oz  \
name                                                                                         
100:Complete First Season (blu-Ray)                                                    0.0   
2x Ultra Era with Oxi Booster, 50fl oz                                                 1.0   
4C Grated Parmesan Cheese 100% Natural 8oz Shaker                                      0.0   
Africa's Best No-Lye Dual Conditioning Relaxer ...                                     0.0   
Alex Cross (dvdvideo)                                                                  0.0   

                                                    4C Grated Parmesan Cheese 100% Natural 8oz Shaker  \
name                                                                                                    
100:Complete First Season (blu-Ray)                                                               0.0   
2x Ultra Era with Oxi Booster, 50fl oz                                                            0.0   
4C Grated Parmesan Cheese 100% Natural 8oz Shaker                                                 0.0   
Africa's Best No-Lye Dual Conditioning Relaxer ...                                                0.0   
Alex Cross (dvdvideo)                                                                             0.0   

                                                    Africa's Best No-Lye Dual Conditioning Relaxer System Super  \
name                                                                                                              
100:Complete First Season (blu-Ray)                                                               0.0             
2x Ultra Era with Oxi Booster, 50fl oz                                                            0.0             
4C Grated Parmesan Cheese 100% Natural 8oz Shaker                                                 0.0             
Africa's Best No-Lye Dual Conditioning Relaxer ...                                                1.0             
Alex Cross (dvdvideo)                                                                             0.0             

                                                    Alex Cross (dvdvideo)  \
name                                                                        
100:Complete First Season (blu-Ray)                                   0.0   
2x Ultra Era with Oxi Booster, 50fl oz                                0.0   
4C Grated Parmesan Cheese 100% Natural 8oz Shaker                     0.0   
Africa's Best No-Lye Dual Conditioning Relaxer ...                    0.0   
Alex Cross (dvdvideo)                                                 1.0   

                                                    All,bran Complete Wheat Flakes, 18 Oz.  \
name                                                                                         
100:Complete First Season (blu-Ray)                                                    0.0   
2x Ultra Era with Oxi Booster, 50fl oz                                                 0.0   
4C Grated Parmesan Cheese 100% Natural 8oz Shaker                                      0.0   
Africa's Best No-Lye Dual Conditioning Relaxer ...                                     0.0   
Alex Cross (dvdvideo)                                                                  0.0   

              

In [67]:
common_item_based_matrix = common.pivot_table(index='reviews_username', columns='name', values='reviews_rating')

In [68]:
common_item_based_matrix.shape

(7926, 196)

In [69]:
item_correlation_df_3[item_correlation_df_3<0]=0

common_item_predicted_ratings = np.dot(item_correlation_df_3, common_item_based_matrix.T.fillna(0))
common_item_predicted_ratings

array([[0.        , 0.        , 0.        , ..., 0.01242535, 0.03266576,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.        , 0.        , 0.        , ..., 0.00660728, 0.        ,
        0.00032046],
       [0.00799822, 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.01314848, 0.        ,
        0.        ]])

In [70]:
common_item_predicted_ratings.shape

(196, 7926)

In [71]:
dummy_test = common.copy()
dummy_test['reviews_rating'] = dummy_test['reviews_rating'].apply(lambda x: 1 if x>=1 else 0)
dummy_test = dummy_test.pivot_table(index='reviews_username', columns='name', values='reviews_rating').T.fillna(0)
common_item_predicted_ratings = np.multiply(common_item_predicted_ratings,dummy_test)
common_item_predicted_ratings.shape

(196, 7926)

In [72]:
common_ = common.pivot_table(index='reviews_username', columns='name', values='reviews_rating').T

In [73]:
from sklearn.preprocessing import MinMaxScaler
from numpy import *

X  = common_item_predicted_ratings.copy() 
X = X[X>0]

scaler = MinMaxScaler(feature_range=(1, 5))
print(scaler.fit(X))
y = (scaler.transform(X))

print(y)

MinMaxScaler(feature_range=(1, 5))
[[nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 ...
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]]


C:\Users\BA06878\Anaconda3\lib\site-packages\sklearn\preprocessing\_data.py:461: RuntimeWarning: All-NaN slice encountered
  data_min = np.nanmin(X, axis=0)
C:\Users\BA06878\Anaconda3\lib\site-packages\sklearn\preprocessing\_data.py:462: RuntimeWarning: All-NaN slice encountered
  data_max = np.nanmax(X, axis=0)


In [74]:
# Finding total non-NaN value
total_non_nan = np.count_nonzero(~np.isnan(y))

In [75]:
rmse = (sum(sum((common_ - y )**2))/total_non_nan)**0.5
print(rmse)

3.571158589545605


## Pickle the selected file

In [76]:
import pickle

In [79]:
file = './pickle/recommend_data.pkl'
fileObj = open(file, 'wb')
pickle.dump(item_final_rating, fileObj)
fileObj.close()

In [80]:
fileReadObj = open(file, 'rb')
pickleObj = pickle.load(fileReadObj)
pickleObj.shape

(17891, 255)

In [81]:
def recomm(item_rating, user_id):
    return item_rating.loc[user_id].sort_values(ascending=False)[0:20]

In [84]:
recommend_file = './pickle/recommendFn.pkl'
fileObj1 = open(recommend_file, 'wb')
pickle.dump(recomm, fileObj1)
fileObj.close()

In [87]:
fileReadObj1 = open(recommend_file, 'rb')
recommend = pickle.load(fileReadObj1)
val = recommend(item_final_rating, '00sab00')

In [88]:
list(val.index)[0]

'Jolly Time Select Premium Yellow Pop Corn'

In [89]:
val

name
Jolly Time Select Premium Yellow Pop Corn                                                        0.055762
Bilbao Nightstand Gray Oak - South Shore                                                         0.050248
Feit 60-Watt A19 Gu24 Base Led Light Bulb - Soft White                                           0.045734
Godzilla 3d Includes Digital Copy Ultraviolet 3d/2d Blu-Ray/dvd                                  0.023402
100:Complete First Season (blu-Ray)                                                              0.017037
Hoover174 Platinum Collection153 Lightweight Bagged Upright Vacuum With Canister - Uh30010com    0.010978
Jason Aldean - They Don't Know                                                                   0.010718
The Resident Evil Collection 5 Discs (blu-Ray)                                                   0.010279
Musselman Apple Sauce, Cinnamon, 48oz                                                            0.007999
Ogx Conditioner, Hydrating Teatree Mint  

# Summary

- rmse for Item Based recommendation: 3.571158589545605
- rmse for User Based recommendation: 2.5597195521751623

Though both rmase is not upto mark here. But User Based recommendation is better. So we have pickled used Based recommendation prediction